### PR TITLE
Convert to markdown: fsdp.rst, func.api.rst, func.batch_norm.rst, func.migrating.rst, func.rst 

### LINKED ISSUE #155021: Convert to markdown: fsdp.rst, func.api.rst, func.batch_norm.rst, func.migrating.rst, func.rst
Convert the following files to MyST markdown and ensure that the doc tests are green:

- [fsdp.rst](https://github.com/pytorch/pytorch/tree/main/docs/source/fsdp.rst)
- [func.api.rst](https://github.com/pytorch/pytorch/tree/main/docs/source/func.api.rst)
- [func.batch_norm.rst](https://github.com/pytorch/pytorch/tree/main/docs/source/func.batch_norm.rst)
- [func.migrating.rst](https://github.com/pytorch/pytorch/tree/main/docs/source/func.migrating.rst)
- [func.rst](https://github.com/pytorch/pytorch/tree/main/docs/source/func.rst)

To convert an `.rst` file:

1) Assign the issue to yourself using the `/assigntome` comment.

2) **Important!** Use `git mv filename.rst filename.md` to convert the issue from `.rst` to `.md`. **DO NOT delete**
   the `.rst` and create the `.md` - we won't accept PRs with deleted `.rst` files. We want to preserve the git log of
   those files. Therefore, using `git mv` is very important!

3) Before submitting a PR, build the docs locally as described in the [README](https://github.com/pytorch/pytorch/blob/main/README.md#building-the-documentation).
   Preview localy with `make serve` to ensure that the HTML renders correctly.    

4) Run `lintrunner` from the root of the repository:

   ```
   pip install lintrunner
   lintrunner init
   lintrunner -m main
   ```

4) Submit a PR with your changes and ensure the PR passes the doc CI tests. Specifically it needs to pass:

   * pull / linux-docs / build-docs-python-false (pull_request)
   * linux-jammy-py3.9-gcc11 / test (docs_test, 1, 1, lf.linux.2xlarge)

   Doc tests start after the Docker build, which might take about 30 minutes. 
   The doc build itself should take about 25 minutes. When the build is done, 
   you can see the preview of your build by clicking:
   📄 **Preview Python docs built from this PR** in the `pytorch-bot` comment.

   Make sure your PR is tagged with `docathon-h1-2025`, `module: docs`, `topic: not user facing`,
   and `topic: docs` labels.

5) Compare the preview of the pages you converted with same pages it the main branch.
   The converted pages in preview must look the same as in the main branch.

6) Once the PR is accepted, one of the maintainers will merge it.

See [Quick MyST Conversion Reference](https://github.com/pytorch/pytorch/wiki/Docstring-Guidelines#restructuredtext-vs-markdown)
in our wiki or [MyST Parser documentation](https://myst-parser.readthedocs.io/en/v0.15.1/sphinx/use.html) for detailed syntax.

Here is an example [PR](https://github.com/pytorch/pytorch/pull/154438)

cc @sekyondaMeta @AlannaBurke

### LINKED ISSUE #155158: ERROR: Unknown target name: "deepspeed"
### 📚 The doc issue

while building the docs I found the following issue:

```
/localdisk/jafraust/miniforge/envs/docs/lib/python3.10/site-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py:docstring of torch.distributed.fsdp.fully_sharded_data_parallel.FullyShardedDataParallel:3: ERROR: Unknown target name: "xu et al.".
/localdisk/jafraust/miniforge/envs/docs/lib/python3.10/site-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py:docstring of torch.distributed.fsdp.fully_sharded_data_parallel.FullyShardedDataParallel:3: ERROR: Unknown target name: "deepspeed".
looking for now-outdated files... none found
```

### Suggest a potential alternative/fix

change the format from this

```
    This is inspired by `Xu et al.`_ as well as the ZeRO Stage 3 from DeepSpeed_.
    FullyShardedDataParallel is commonly shortened to FSDP.

    .. _`Xu et al.`: https://arxiv.org/abs/2004.13336
    .. _DeepSpeed: https://www.deepspeed.ai/
````

to

```
his is inspired by `Xu et al. <https://arxiv.org/abs/2004.13336>`_ as well as the ZeRO Stage 3 from `DeepSpeed <https://www.deepspeed.ai/>`_.
    FullyShardedDataParallel is commonly shortened to FSDP.
```

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @svekars @sekyondaMeta @AlannaBurke

### PR DESCRIPTION
**Pull request by jafraustro**
Docathon: 

- Convert to markdown: fsdp.rst, func.api.rst, func.batch_norm.rst, func.migrating.rst, func.rst 

Fixes #155021 
Fixes #155158 

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @svekars @sekyondaMeta @AlannaBurke @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @voznesenskym @penguinwu @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

### CODE REVIEW

**Review by zxiiro — CHANGES_REQUESTED** [review]
Looks like this PR only adds 5 new files. I think the intent of of this task to is rename the rst files to md so we'd need another commit to remove the 5 corresponding rst files.

**Comment by sekyondaMeta** [comment]
Double check the issue for instructions on how to correctly convert the .rst to .md (step 2 in the issue gives you the correct way to do it).

**Comment by jafraustro** [comment]
Sorry my bad.

I made the changes

**Review by zxiiro — DISMISSED** [review]

**Comment by jafraustro** [comment]
This PR also address #155158

File: docs/source/func.api.md [hunk 1]
```diff
@@ -21,67 +24,71 @@ Function Transforms
...
+x = torch.randn(3)
+jacobian = jacrev(f)(x)
+assert jacobian.shape == (3, 3)
+```
 
 However, if you want to do something like compute a jacobian over the parameters
 of the model, then there needs to be a way to construct a function where the
 parameters are the inputs to the function.
-That's what :func:`functional_call` is for:
+That's what {func}`functional_call` is for:
```
Comment (svekars) [inline docs/source/func.api.md]: below `parameters` should have single backticks

File: docs/source/func.migrating.md [hunk 2]
```diff
@@ -1,148 +1,143 @@
...
-functorch.jvp                   :func:`torch.func.jvp`
-functorch.jacrev                :func:`torch.func.jacrev`
-functorch.jacfwd                :func:`torch.func.jacfwd`
-functorch.hessian               :func:`torch.func.hessian`
-functorch.functionalize         :func:`torch.func.functionalize`
-==============================  =======================================
+| functorch API                      | PyTorch API (as of PyTorch 2.0)                |
+| ----------------------------------- | ---------------------------------------------- |
+| functorch.vmap                      | `torch.vmap` or `torch.func.vmap`              |
+| functorch.grad                      | `torch.func.grad`                              |
```
Comment (svekars) [inline docs/source/func.migrating.md]: we should use {func}

**Comment by svekars** [comment]
We typically prefer each PR to address one issue but given what a small change this is, I guess it's okay.

File: docs/source/func.api.md (hunk 1 above)
Comment (jafraustro) [reply docs/source/func.api.md]: Fixed

File: docs/source/func.migrating.md (hunk 2 above)
Comment (jafraustro) [reply docs/source/func.migrating.md]: Fixed

File: docs/source/func.migrating.md [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+### functorch.combine_state_for_ensemble
+
+Please use {func}`torch.func.stack_module_state` instead of
+[functorch.combine_state_for_ensemble](https://pytorch.org/functorch/1.13/generated/functorch.combine_state_for_ensemble.html)
+{func}`torch.func.stack_module_state` returns two dictionaries, one of stacked parameters, and
+one of stacked buffers, that can then be used with {func}`torch.vmap` and {func}`torch.func.functional_call`
+for ensembling.
+
+For example, here is an example of how to ensemble over a very simple model::
+
```
Comment (svekars) [inline docs/source/func.migrating.md]:
```suggestion

```python
```

File: docs/source/func.migrating.md [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+### functorch.combine_state_for_ensemble
+
+Please use {func}`torch.func.stack_module_state` instead of
+[functorch.combine_state_for_ensemble](https://pytorch.org/functorch/1.13/generated/functorch.combine_state_for_ensemble.html)
+{func}`torch.func.stack_module_state` returns two dictionaries, one of stacked parameters, and
+one of stacked buffers, that can then be used with {func}`torch.vmap` and {func}`torch.func.functional_call`
+for ensembling.
+
+For example, here is an example of how to ensemble over a very simple model::
```
Comment (svekars) [inline docs/source/func.migrating.md]:
```suggestion
For example, here is an example of how to ensemble over a very simple model:
```

File: docs/source/func.migrating.md [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+    params, buffers = torch.func.stack_module_state(models)
+
+    # It is possible to vmap directly over torch.func.functional_call,
+    # but wrapping it in a function makes it clearer what is going on.
+    def call_single_model(params, buffers, data):
+        return torch.func.functional_call(base_model, (params, buffers), (data,))
+
+    output = torch.vmap(call_single_model, (0, 0, None))(params, buffers, data)
+    assert output.shape == (num_models, batch_size, out_features)
```
Comment (svekars) [inline docs/source/func.migrating.md]: Need to add the closing markdown backticks.

File: docs/source/func.migrating.md [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+
+Please use {func}`torch.func.stack_module_state` instead of
+[functorch.combine_state_for_ensemble](https://pytorch.org/functorch/1.13/generated/functorch.combine_state_for_ensemble.html)
+{func}`torch.func.stack_module_state` returns two dictionaries, one of stacked parameters, and
+one of stacked buffers, that can then be used with {func}`torch.vmap` and {func}`torch.func.functional_call`
+for ensembling.
+
+For example, here is an example of how to ensemble over a very simple model::
+
+    import torch
```
Comment (svekars) [inline docs/source/func.migrating.md]: please: indent to the left the whole block

The formatting is a bit different: https://docs-preview.pytorch.org/pytorch/pytorch/155104/func.migrating.html#functorch-combine-state-for-ensemble from the main: https://docs.pytorch.org/docs/main/func.migrating.html#functorch-combine-state-for-ensemble

**Review by svekars — COMMENTED** [review]
Looks good - I've found a couple of minor formatting issues in func.migrating.md but otherwise this will be good to go.

**Comment by jafraustro** [comment]
Thanks for your help @svekars

**Review by svekars — APPROVED** [review]

**Comment by svekars** [comment]
Can you please make sure to update/rebase locally to main and and update the PR? It will be ready to merge after that.

**Comment by jafraustro** [comment]
> [refs/remotes/origin/viable/strict](https://github.com/pytorch/pytorch/commit/e895e9689c625cbcd8f46880115e0d093713fa37).

I made the rebase onto [refs/remotes/origin/viable/strict](https://github.com/pytorch/pytorch/commit/e895e9689c625cbcd8f46880115e0d093713fa37) locally.

**Comment by svekars** [comment]
SOmething went wrong with the rebase and i see now 182 files changed.

**Review by zou3519 — CHANGES_REQUESTED** [review]
why do you need to change .cpp files when migrating .rst to markdown?

**Comment by jafraustro** [comment]
> SOmething went wrong with the rebase and i see now 182 files changed.
Yes, I made the rebase against viable/strict instead of main. I am closing this PR and continue in #155369
